ai_powered_qa/components/agent.py

```diff
@@ -153,23 +153,26 @@ def _get_messages_for_completion(
         if user_prompt:
             total_tokens += count_tokens(user_prompt, model)
 
-        for i in range(len(self.history)):
+        messages_to_add = []
+        i = 0
+
+        while i < len(self.history):
             history_item = self.history[-i - 1]
-            content_length = count_tokens(history_item["content"], model)
+            messages_to_add.insert(0, history_item)
+            content_length = count_tokens(yaml.dump(history_item), model)
```

Comment:
I'm dumping the whole messages into yaml. There's some token overhead for each message anyway, and it makes the code a bit cleaner.